src/datasets/table.py

```diff
@@ -0,0 +1,591 @@
+
+    @wraps_arrow_table_method(pa.Table.set_column)
+    def set_column(self, *args, **kwargs):
+        replay = [("set_column", copy.deepcopy(args), copy.deepcopy(kwargs))]
+        replay = self._merge_replay(replay)
+        return MemoryMappedTable(self.table.set_column(*args, **kwargs), self.path, replay)
+
+    @wraps_arrow_table_method(pa.Table.rename_columns)
+    def rename_columns(self, *args, **kwargs):
+        replay = [("rename_columns", copy.deepcopy(args), copy.deepcopy(kwargs))]
+        replay = self._merge_replay(replay)
+        return MemoryMappedTable(self.table.rename_columns(*args, **kwargs), self.path, replay)
+
+    @wraps_arrow_table_method(pa.Table.drop)
+    def drop(self, *args, **kwargs):
+        replay = [("drop", copy.deepcopy(args), copy.deepcopy(kwargs))]
+        replay = self._merge_replay(replay)
+        return MemoryMappedTable(self.table.drop(*args, **kwargs), self.path, replay)
+
+
+BlockPart = TypeVar("BlockPart", TableBlock, List[TableBlock], List[List[TableBlock]])
+
+
+class ConcatenationTable(Table):
+    """
+    Wraps a pyarrow Table using composition.
+    The table comes from the concatenation of several tables.
+
+    The underlying tables are called "blocks" and can be either InMemoryTable
+    or MemoryMappedTable objects.
+    This allows to combine tables that come from memory or that are memory mapped.
+    When a ConcatenationTable is pickled, then each block is pickled:
+    - the InMemoryTable objects are pickled by copying all the data in memory;
+    - the MemoryMappedTable objects are pickled without copying the data into memory.
+    Instead, only the path to the memory mapped arrow file is pickled, as well as the list
+    of transforms to "replay" when reloading the table from the disk.
+
+    Its implementation requires to store each block separately.
+    The ``blocks`` attributes stores a list of list of blocks.
+    The first axis concatenates the tables along the axis 0 (it appends rows),
+    while the second axis concatenates tables along the axis 1 (it appends columns).
+
+    You can access the fully combined table by accessing the ConcatenationTable.table attribute,
+    and the blocks by accessing the ConcatenationTable.blocks attribute.
+    """
+
+    def __init__(self, table: pa.Table, blocks: List[List[TableBlock]]):
+        super().__init__(table)
+        self.blocks = blocks
+        assert all(isinstance(subtable, TableBlock) for subtables in blocks for subtable in subtables)
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state.pop("table")
+        return state
+
+    def __setstate__(self, state):
+        state = state.copy()
+        state["table"] = ConcatenationTable.from_blocks(state["blocks"]).table
+        self.__dict__ = state
+
+    @staticmethod
+    def from_blocks(blocks_part: BlockPart):
+        if isinstance(blocks_part, Table):
```

Comment:
Why not checking more specifically that is instance of `TableBlock`?

The same below.